File: rust/crates/agent-daemon/src/workspaces/local.rs
use std::path::{Path, PathBuf};

use anyhow::{bail, Context, Result};
use tokio::process::Command;

use super::config::{required_local_field, WorkspaceBaseConfig};
use super::sanitize::sanitize_copied_tree;

pub(super) async fn refresh_local_workspace_base(
    base: &Path,
    config: &WorkspaceBaseConfig,
) -> Result<()> {
    let source_path = required_local_field(config.source_path.as_deref(), "source_path")?;
    let source = PathBuf::from(source_path);
    if !source.is_dir() {
        bail!(
            "local workspace source_path is not a directory: {}",
            source.display()
        );
    }
    rsync_dir_delete(&source, base).await?;
    sanitize_copied_tree(base).await
}

async fn rsync_dir_delete(source: &Path, target: &Path) -> Result<()> {
    let source_contents = source.join(".");
    let output = Command::new("rsync")
        .arg("-a")
        .arg("--delete")
        .arg("--delete-excluded")
        .arg("--numeric-ids")
        .arg("--no-owner")
        .arg("--no-group")
        .arg(source_contents)
        .arg(target)
        .output()
        .await
        .with_context(|| {
            format!(
                "rsync local workspace base {} to {}",
                source.display(),
                target.display()
            )
        })?;
    if !output.status.success() {
        bail!(
            "rsync failed from {} to {}: {}",
            source.display(),
            target.display(),
            String::from_utf8_lossy(&output.stderr).trim()
        );
    }
    Ok(())
}

File: packages/workspace-lib/src/local.ts
// Local workspace base refresh (port of workspaces/local.rs) + tree sanitize
// (port of sanitize.rs: strip absolute/escaping symlinks, drop special files).
import { lstatSync, readdirSync, readlinkSync, rmSync, writeFileSync } from "node:fs";
import { join, isAbsolute } from "node:path";
import { run } from "./exec.ts";
import type { WorkspaceBaseConfig } from "./config.ts";

export async function refreshLocalWorkspaceBase(base: string, config: WorkspaceBaseConfig): Promise<void> {
	const source = config.source_path ?? "";
	// rsync -a --delete --delete-excluded --numeric-ids --no-owner --no-group src/. dst
	await run("rsync", ["-a", "--delete", "--delete-excluded", "--numeric-ids", "--no-owner", "--no-group", `${source}/.`, base]);
	sanitizeCopiedTree(base);
}

function isSafeRelativeSymlink(target: string): boolean {
	if (isAbsolute(target)) return false;
	for (const part of target.split("/")) {
		if (part === "..") return false;
		if (part === "" || part === ".") continue;
	}
	return true;
}

function writeSkippedSymlinkMarker(path: string, target: string): void {
	writeFileSync(path, `pi-relay local workspace copy skipped external symlink target: ${target}\n`);
}

export function sanitizeCopiedTree(dir: string): void {
	for (const entry of readdirSync(dir, { withFileTypes: true })) {
		const child = join(dir, entry.name);
		if (entry.isDirectory()) {
			sanitizeCopiedTree(child);
		} else if (entry.isSymbolicLink()) {
			const target = readlinkSync(child);
			if (!isSafeRelativeSymlink(target)) {
				rmSync(child, { force: true });
				writeSkippedSymlinkMarker(child, target);
			}
		} else if (!entry.isFile()) {
			rmSync(child, { force: true, recursive: true });
		}
	}
}
